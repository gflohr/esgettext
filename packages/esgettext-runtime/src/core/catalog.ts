export interface CatalogEntries {
	[key: string]: Array<string>;
}

export interface Catalog {
	major: number;
	minor: number;
	pluralFunction: Function;
	entries: CatalogEntries;
}
