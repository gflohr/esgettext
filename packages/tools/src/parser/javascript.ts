import { decode } from 'iconv-lite';
import { parse } from '@babel/parser';
import { Parser } from './parser';

export class JavaScriptParser extends Parser {
	parse(buf: Buffer, filename: string, encoding?: string): boolean {
		const input =
			typeof encoding === 'undefined' ? buf.toString() : decode(buf, encoding);

		const ast = parse(input, {
			allowAwaitOutsideFunction: true,
			allowImportExportEverywhere: true,
			allowReturnOutsideFunction: true,
			allowSuperOutsideMethod: true,
			allowUndeclaredExports: true,
			// Documented but not supported. FIXME! Maybe only missing in types.
			// errorRecovery: true,
			sourceFilename: filename,
			plugins: ['flow'],
		});

		return this.extract(filename, ast);
	}
}
