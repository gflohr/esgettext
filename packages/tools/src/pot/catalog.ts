import { POTEntry } from './entry';

interface Cache {
	// msgid
	[key: string]: {
		// msgctxt
		[key: string]: POTEntry;
	};
}

interface CatalogProperties {
	package?: string;
	version?: string;
	msgidBugsAddress?: string;
	foreignUser?: boolean;
	date?: string;
	fromCode?: string;
	sortOutput?: boolean;
	sortByFile?: boolean;
}

interface RenderOptions {
	width?: number;
}

/**
 * An entire message catalog that is subsequently filled.
 */
export class Catalog {
	private readonly cache: Cache = {};
	private readonly entries: Array<POTEntry>;

	constructor(private readonly properties: CatalogProperties = {}) {
		if (typeof properties.package === 'undefined') {
			properties.package = 'PACKAGE';
		}
		if (typeof properties.version === 'undefined') {
			properties.version = 'VERSION';
		}
		if (typeof properties.msgidBugsAddress === 'undefined') {
			properties.msgidBugsAddress = 'MSGID_BUGS_ADDRESS';
		}
		// GNU xgettext does not set the charset in the PO header but we do.
		if (typeof properties.fromCode === 'undefined') {
			properties.fromCode = 'CHARSET';
		}
		if (typeof properties.date === 'undefined') {
			const now = new Date();

			// Avoid if/else, so we do not spoil our test coverage. :)
			let year = now.getFullYear().toString();
			year = '0'.repeat(4 - year.length) + year;
			let month = now.getMonth().toString();
			month = '0'.repeat(2 - month.length) + month;
			let mday = now.getDate().toString();
			mday = '0'.repeat(2 - mday.length) + mday;
			let hour = now.getHours().toString();
			hour = '0'.repeat(2 - hour.length) + hour;
			let minutes = now.getMinutes().toString();
			minutes = '0'.repeat(2 - minutes.length) + minutes;

			// Do not depend on the timezone for test coverage.
			let offset = now.getTimezoneOffset();
			const sign = ['+', '-'][Math.sign(offset) + 1];
			offset = Math.abs(offset);

			let offsetHours = Math.floor(offset / 60).toString();
			offsetHours = '0'.repeat(2 - offsetHours.length) + offsetHours;
			let offsetMinutes = (offset % 60).toString();
			offsetMinutes = '0'.repeat(2 - offsetMinutes.length) + offsetMinutes;

			properties.date =
				[year, month, mday].join('-') +
				` ${hour}:${minutes}` +
				`${sign}${offsetHours}${offsetMinutes}`;
		}

		const pkg = properties.package.replace(/\n/g, '\\n');
		const version = properties.version.replace(/\n/g, '\\n');
		const msgidBugsAddress = properties.msgidBugsAddress.replace(/\n/g, '\\n');

		const header = `Project-Id-Version: ${pkg} ${version}
Report-Msgid-Bugs-To: ${msgidBugsAddress}
POT-Creation-Date: ${properties.date}
PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE
Last-Translator: FULL NAME <EMAIL@ADDRESS>
Language-Team: LANGUAGE <LL@li.org>
Language:
MIME-Version: 1.0
Content-Type: text/plain; charset=${properties.fromCode}
Content-Transfer-Encoding: 8bit
`;

		let comment;
		if (properties.foreignUser) {
			comment = `SOME DESCRIPTIVE TITLE
This file is put in the public domain.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		} else {
			comment = `SOME DESCRIPTIVE TITLE
Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
This file is distributed under the same license as the PACKAGE package.
FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
`;
		}

		this.entries = new Array<POTEntry>();
		const headerEntry = new POTEntry({
			msgid: '',
			msgstr: header,
			flags: ['fuzzy'],
			translatorComments: [comment],
			noWarnings: true,
		});
		this.addEntry(headerEntry);
	}

	/**
	 * Render the entire catalog.
	 *
	 * @param renderOptions - options for rendering
	 *
	 * The following options are supported:
	 *
	 * width: wrap lines to that page width (default: 76)
	 *
	 * @returns the catalog rendered as a po(t) file
	 */
	toString(options?: RenderOptions): string {
		const width =
			options && typeof options.width !== 'undefined' ? options.width : 76;

		if (this.properties.sortOutput) {
			return this.entries
				.sort((a, b) => {
					const cmp = a.properties.msgid.localeCompare(b.properties.msgid);
					if (cmp) {
						return cmp;
					}
					const actx = a.properties.msgctxt;
					const bctx = b.properties.msgctxt;

					if (typeof actx === 'undefined') {
						if (typeof bctx === 'undefined') {
							return 0;
						} else {
							return -1;
						}
					} else {
						if (typeof bctx === 'undefined') {
							return +1;
						} else {
							return actx.localeCompare(bctx);
						}
					}
				})
				.map((entry) => entry.toString(width))
				.join('\n');
		} else {
			return this.entries.map((entry) => entry.toString(width)).join('\n');
		}
	}

	/**
	 * Add an entry to the catalog. If an entry with the same msgid
	 * and msgctxt already exists, the entry is merged into
	 * the existing one instead.
	 *
	 * @param entry - the `POTEntry` to add
	 */
	addEntry(entry: POTEntry): void {
		const msgid = entry.properties.msgid;
		const msgctxt = entry.properties.msgctxt;

		if (!Object.prototype.hasOwnProperty.call(this.cache, msgid)) {
			this.cache[msgid] = {};
		}
		if (!Object.prototype.hasOwnProperty.call(this.cache[msgid], msgctxt)) {
			this.cache[msgid][msgctxt] = entry;
			this.entries.push(entry);
		} else {
			this.cache[msgid][msgctxt].merge(entry);
		}
	}

	/**
	 * Copy a catalog with other options. This is for testing only.
	 *
	 * @returns a deep copy of the catalog.
	 */
	copy(properties?: CatalogProperties): Catalog {
		const other = new Catalog(properties);

		this.entries.forEach((entry) => other.addEntry(entry));

		return other;
	}
}
