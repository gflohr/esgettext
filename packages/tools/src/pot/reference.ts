export class Reference {
	constructor(
		private readonly filename: string,
		private readonly lineNumber: number,
	) {}

	toString() {
		const filename = this.filename.replace('\n', '\\n');

		return `${filename}:${this.lineNumber}`;
	}
}
