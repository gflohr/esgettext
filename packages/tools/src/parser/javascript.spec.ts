import { Catalog } from '../pot/catalog';
import { Keyword } from '../pot/keyword';
import { JavaScriptParser } from './javascript';

const warnSpy = jest.spyOn(global.console, 'warn').mockImplementation(() => {
	/* ignore */
});
const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {
	/* ignore */
});

describe('JavaScript parser', () => {
	describe('strings', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should extract all kinds of strings', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, { extractAll: true });
			const code = `
// Directive.
_("double-quoted string");

// Function call.
_('single-quoted string');

// perl-brace-format.
_x('Hello, {name}!');
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toMatchSnapshot();
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract concatenated strings', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, { extractAll: true });
			const code = '"concatenated" + " str" + "i" + "n" + "g";';
			const buf = Buffer.from(code);
			expect(p.parse(buf, 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "concatenated string"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract concatenated strings with a leading variable', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, { extractAll: true });
			const code = 'prefix + "concatenated" + " string";';
			const buf = Buffer.from(code);
			expect(p.parse(buf, 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract concatenated strings mixed with a variable', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, { extractAll: true });
			const code = '"concatenated" + sep + "string";';
			const buf = Buffer.from(code);
			expect(p.parse(buf, 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract concatenated strings with a trailing variable', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, { extractAll: true });
			const code = '"concatenated" + " string" + suffix;';
			const buf = Buffer.from(code);
			expect(p.parse(buf, 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});

	describe('comments', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should extract all kinds of comments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				extractAll: true,
				addAllComments: true,
			});
			const code = `
// Single-line comment.
_('single line');

/* Block comment. */
_('block');

// Single-line
/* and block combined. */
_('single + block');

/* Block */
// and single-line combined.
_('block + single');

// Skip this.

// But catch this.
_('only immediately preceding');

// Only for one string.
_('catcher'); _('loser');

// And this is ignored, too.
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toMatchSnapshot();
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});

	describe('messages', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should extract a single argument', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("Hello, world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "Hello, world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract singular and plural form', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_n', ['1', '2'])],
			});
			const code = '_n("There was an error!", "There were errors!", n)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "There was an error!"
msgid_plural "There were errors!"
msgstr[0] ""
msgstr[1] ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract context, singular and plural form', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code =
				'_np("context", "There was an error!", "There were errors!", n)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgctxt "context"
msgid "There was an error!"
msgid_plural "There were errors!"
msgstr[0] ""
msgstr[1] ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should handle other callees', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '"string".trim()';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract from unknown methods', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = 'gettext("string")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should honor the total arguments spec', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1t'])],
			});
			const code = '_("Hello", "world")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract non existing singular arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_()';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract non existing plural arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_n', ['1', '2'])],
			});
			const code = '_n("One universe")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should not extract non existing context arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_pn', ['1', '2', '3c'])],
			});
			const code = '_pn("one world", "many worlds")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject variables as singular arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", earth, "many earths")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should allow simple template literals as singular arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", `one earth`, "many earths")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgctxt "world"
msgid "one earth"
msgid_plural "many earths"
msgstr[0] ""
msgstr[1] ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject template literals with embedded expressions as singular arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", `one ${planet}`, "many earths")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeFalsy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).toHaveBeenCalledTimes(1);
			expect(errorSpy).toHaveBeenNthCalledWith(
				1,
				'example.js:1:13-1:28: Error: template literals with embedded' +
					' expressions are not allowed as arguments to gettext' +
					' functions because they are not constant',
			);
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject variables as plural arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", "earth", earths)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should allow simple template literals as plural arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", "one earth", `many earths`)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgctxt "world"
msgid "one earth"
msgid_plural "many earths"
msgstr[0] ""
msgstr[1] ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject template literals with embedded expressions as plural arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", "one earth", `many ${planets}`)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeFalsy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).toHaveBeenCalledTimes(1);
			expect(errorSpy).toHaveBeenNthCalledWith(
				1,
				'example.js:1:26-1:43: Error: template literals with embedded' +
					' expressions are not allowed as arguments to gettext' +
					' functions because they are not constant',
			);
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject variables as msgctxt arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np(world, "earth", "many earths")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should allow simple template literals as msgctxt arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np(`world`, "one earth", "many earths")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgctxt "world"
msgid "one earth"
msgid_plural "many earths"
msgstr[0] ""
msgstr[1] ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should reject template literals with embedded expressions as msgctxt arguments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_np', ['1c', '2', '3'])],
			});
			const code = '_np("world", "one earth", `many ${planets}`)';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeFalsy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).toHaveBeenCalledTimes(1);
			expect(errorSpy).toHaveBeenNthCalledWith(
				1,
				'example.js:1:26-1:43: Error: template literals with embedded' +
					' expressions are not allowed as arguments to gettext' +
					' functions because they are not constant',
			);
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should respect the selected instance names', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1'])],
				instance: ['gtx'],
			});
			const code = 'gtx._("gotcha!");\nother._("no, no!");';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "gotcha!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract instances inside this expressions', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1'])],
			});
			const code = 'this.gtx._("gotcha!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "gotcha!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract computed method calls', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1'])],
			});
			const code = 'gtx["_"]("gotcha!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "gotcha!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract computed method calls inside this expressions', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1'])],
			});
			const code = 'this["_"]("gotcha!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "gotcha!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract calls inside template strings', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_', ['1'])],
			});
			const code = '`Hello, ${gtx._("Hello, world!")}`';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "Hello, world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});

	describe('encoding', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should decode to JavaScript strings', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("Hello, world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "Hello, world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});

	describe('binary expressions', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should extract concatenated strings', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("It\'s a " + "sad" + " and beautiful world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "It's a sad and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract concatenated strings with simple template literals', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("It\'s a " + `sad` + " and beautiful world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:1
msgid "It's a sad and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should ignore concatenated strings with variables', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("It\'s a " + what + " and beautiful world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			expect(catalog.toString({ omitHeader: true })).toEqual('');
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should report concatenated strings with interpolations', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = '_("It\'s " + `a ${what}` + " and beautiful world!")';
			expect(p.parse(Buffer.from(code), 'example.js')).toBeFalsy();
			expect(errorSpy).toHaveBeenCalledTimes(1);
			expect(errorSpy).toHaveBeenCalledWith(
				'example.js:1:12-1:23: Error:' +
					' template literals with embedded expressions are not' +
					' allowed as arguments to gettext functions because' +
					' they are not constant',
			);
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});

	describe('comments above calls', () => {
		afterEach(() => {
			warnSpy.mockClear();
			errorSpy.mockClear();
		});

		it('should extract xgettext: comments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = `// xgettext: no-perl-brace-format
_("It's a {sad} and beautiful world!")
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:2
#, no-perl-brace-format
msgid "It's a {sad} and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should ignore invalid xgettext: comments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
			});
			const code = `// xgettext: down-by-law
_("It's a sad and beautiful world!")
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:2
msgid "It's a sad and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should extract translator comments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
				addComments: ['TRANSLATORS:'],
			});
			const code = `// TRANSLATORS: Down by Law
_("It's a sad and beautiful world!")
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#. TRANSLATORS: Down by Law
#: example.js:2
msgid "It's a sad and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});

		it('should ignore other comments', () => {
			const catalog = new Catalog();
			const p = new JavaScriptParser(catalog, {
				keyword: [new Keyword('_')],
				addComments: ['TRANSLATORS:'],
			});
			const code = `// TERMINATORS: Down by Law
_("It's a sad and beautiful world!")
`;
			expect(p.parse(Buffer.from(code), 'example.js')).toBeTruthy();
			const expected = `#: example.js:2
msgid "It's a sad and beautiful world!"
msgstr ""
`;
			expect(catalog.toString({ omitHeader: true })).toEqual(expected);
			expect(errorSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
			expect(warnSpy).not.toHaveBeenCalled();
		});
	});
});
