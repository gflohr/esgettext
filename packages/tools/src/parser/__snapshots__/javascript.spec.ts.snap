// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JavaScript parser comments should extract all kinds of comments 1`] = `
"#. Single-line comment.
#: example.ts:3
msgid \\"single line\\"
msgstr \\"\\"

#. Block comment.
#: example.ts:6
msgid \\"block\\"
msgstr \\"\\"

#. Single-line
#. and block combined.
#: example.ts:10
msgid \\"single + block\\"
msgstr \\"\\"

#. Block
#. and single-line combined.
#: example.ts:14
msgid \\"block + single\\"
msgstr \\"\\"

#. But catch this.
#: example.ts:19
msgid \\"only immediately preceding\\"
msgstr \\"\\"

#. Only for one string.
#: example.ts:22
msgid \\"catcher\\"
msgstr \\"\\"

#: example.ts:22
msgid \\"loser\\"
msgstr \\"\\"
"
`;

exports[`JavaScript parser strings should extract all kinds of strings 1`] = `
"#: example.ts:3
msgid \\"double-quoted string\\"
msgstr \\"\\"

#: example.ts:6
msgid \\"single-quoted string\\"
msgstr \\"\\"

#: example.ts:9
#, perl-brace-format
msgid \\"Hello, {name}!\\"
msgstr \\"\\"
"
`;
